wrapper/src/main/java/software/amazon/jdbc/OpenedConnectionTracker.java

```diff
@@ -0,0 +1,163 @@
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc;
+
+import java.lang.ref.WeakReference;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.logging.Logger;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.RdsUrlType;
+import software.amazon.jdbc.util.RdsUtils;
+
+public class OpenedConnectionTracker {
+
+  public static final Map<String, Queue<WeakReference<Connection>>> openedConnections = new ConcurrentHashMap<>();
+  private static final ExecutorService invalidateConnectionsExecutorService =
+      Executors.newCachedThreadPool(
+          r -> {
+            final Thread invalidateThread = new Thread(r);
+            invalidateThread.setDaemon(true);
+            return invalidateThread;
+          });
+  private static final ExecutorService abortConnectionExecutorService =
+      Executors.newCachedThreadPool(
+          r -> {
+            final Thread abortThread = new Thread(r);
+            abortThread.setDaemon(true);
+            return abortThread;
+          });
+
+  private static final Logger LOGGER = Logger.getLogger(OpenedConnectionTracker.class.getName());
+  private static final RdsUtils rdsUtils = new RdsUtils();
+
+  public static void populateOpenedConnectionQueue(final HostSpec hostSpec, final Connection conn) {
+    final Set<String> aliases = hostSpec.asAliases();
+
+    // Check if the connection was established using a cluster endpoint
+    final RdsUrlType urlType = rdsUtils.identifyRdsType(hostSpec.getHost());
+    String endpoint = null;
+    if (urlType.isRdsCluster()) {
+      // 1. Check if aliases contain an instance endpoint.
+      for (String alias : aliases) {
+        if (!rdsUtils.identifyRdsType(alias).isRdsCluster()) {
+          endpoint = alias;
+          break;
+        }
+      }
+
+      if (endpoint == null) {
+        LOGGER.fine(Messages.get("OpenedConnectionTracker.unableToPopulateOpenedConnectionQueue"));
+        return;
+      }
+    }
+
+    Queue<WeakReference<Connection>> connectionQueue = getConnectionQueue(aliases);
+    connectionQueue.add(new WeakReference<>(conn));
+
+    for (String alias : aliases) {
```

Comment:
It seems we don't need this code since connection should always be in the map wit instance endpoint as a key.